controller/konnect/reconciler_generic.go

```diff
@@ -211,6 +211,27 @@ func (r *KonnectEntityReconciler[T, TEnt]) Reconcile(
 		return res, nil
 	}
 
+	// If a type has a KongUpstream ref (KongTarget), handle it.
+	res, err = handleKongUpstreamRef(ctx, r.Client, ent)
+	if err != nil {
+		if !errors.As(err, &ReferencedKongUpstreamIsBeingDeleted{}) {
+			return ctrl.Result{}, err
+		}
+
+		// If the referenced KongUpstream is being deleted (has a non zero deletion timestamp)
+		// then we remove the entity if it has not been deleted yet (deletion timestamp is zero).
+		// We do this because Konnect blocks deletion of entities like Services/Upstreams
+		// if they contain entities like Routes/Targets.
+		if ent.GetDeletionTimestamp().IsZero() {
+			if err := r.Client.Delete(ctx, ent); err != nil {
+				return ctrl.Result{}, fmt.Errorf("failed to delete %s: %w", client.ObjectKeyFromObject(ent), err)
+			}
+			return ctrl.Result{}, nil
+		}
```

Comment:
I've just checked and Konnect allows removing upstreams that contain targets.

Given this, I suggest we implement the same handling as it has been done for consumers and their credentials: https://github.com/Kong/gateway-operator/blob/ca99be3217aab7c788b27d925669aa8250414d09/controller/konnect/reconciler_generic.go#L134-L152

Reply:
Would it run cascade delete on targets if we call API to delete an upstream? If not, it would bring orphaned targets and we cannot delete them.

Reply:
Updated.

Reply:
> Would it run cascade delete on targets if we call API to delete an upstream? If not, it would bring orphaned targets and we cannot delete them.

I can see we're (correctly) setting the owner reference in https://github.com/Kong/gateway-operator/pull/627/files#diff-bb95aba82e21609e440b7242dc888eda4d9b80d8639c6f336d223035ee199d09R99 (as we do for other refs) so this should cascade delete the target.

In any case an envtest based test for this would be nice to add in a separate  PR ( similarly as we have it for other types).